File: src/parser/__tests__/fixtures/ruby/sample2.rb
# sample2.rb - Modules, mixins, and advanced patterns

# Comparable module for comparison operations
module Comparable
  def ==(other)
    compare_to(other) == 0
  end

  def <(other)
    compare_to(other) < 0
  end

  def >(other)
    compare_to(other) > 0
  end

  def <=(other)
    compare_to(other) <= 0
  end

  def >=(other)
    compare_to(other) >= 0
  end
end

# Drawable module for objects that can be drawn
module Drawable
  def draw
    "Drawing #{self.class.name}"
  end

  def get_bounds
    [0, 0, 0, 0]
  end

  module ClassMethods
    def supported_formats
      [:svg, :png, :pdf]
    end
  end

  def self.included(base)
    base.extend(ClassMethods)
  end
end

# Serializable module
module Serializable
  def to_json
    require 'json'
    instance_variables.each_with_object({}) do |var, hash|
      hash[var.to_s.delete('@')] = instance_variable_get(var)
    end.to_json
  end

  def self.from_json(json_string)
    require 'json'
    data = JSON.parse(json_string)
    instance = new
    data.each do |key, value|
      instance.instance_variable_set("@#{key}", value)
    end
    instance
  end
end

# Container class with generic-like behavior
class Container
  include Serializable

  attr_accessor :value

  def initialize(value = nil)
    @value = value
  end

  def get
    @value
  end

  def set(val)
    @value = val
  end

  def map(&block)
    Container.new(block.call(@value))
  end

  def self.wrap(value)
    new(value)
  end
end

# Pair class holding two values
class Pair
  include Serializable

  attr_reader :first, :second

  def initialize(first, second)
    @first = first
    @second = second
  end

  def swap
    Pair.new(@second, @first)
  end

  def map_first(&block)
    Pair.new(block.call(@first), @second)
  end

  def map_second(&block)
    Pair.new(@first, block.call(@second))
  end
end

# Result class for error handling
class Result
  attr_reader :value, :error

  def initialize(value: nil, error: nil)
    @value = value
    @error = error
  end

  def success?
    @error.nil?
  end

  def failure?
    !success?
  end

  def unwrap
    raise @error if failure?
    @value
  end

  def self.success(value)
    new(value: value)
  end

  def self.failure(error)
    new(error: error)
  end
end

# Shape with Drawable mixin
class Shape
  include Drawable

  attr_reader :color

  def initialize(color)
    @color = color
  end

  def area
    raise NotImplementedError
  end
end

# Rectangle with Drawable
class Rectangle < Shape
  attr_reader :width, :height

  def initialize(color, width, height)
    super(color)
    @width = width
    @height = height
  end

  def area
    @width * @height
  end

  def get_bounds
    [0, 0, @width, @height]
  end
end

# Circle with Drawable
class CircleShape < Shape
  attr_reader :radius

  PI = 3.14159

  def initialize(color, radius)
    super(color)
    @radius = radius
  end

  def area
    PI * @radius ** 2
  end

  def get_bounds
    diameter = 2 * @radius
    [-@radius, -@radius, diameter, diameter]
  end
end

# Observer pattern
module Observable
  def initialize
    super
    @observers = []
  end

  def add_observer(observer)
    @observers << observer unless @observers.include?(observer)
  end

  def remove_observer(observer)
    @observers.delete(observer)
  end

  def notify_observers(event)
    @observers.each { |observer| observer.update(event) }
  end
end

# Subject class using Observable
class Subject
  include Observable

  attr_reader :state

  def initialize
    super
    @state = :idle
  end

  def change_state(new_state)
    @state = new_state
    notify_observers(new_state)
  end
end

# Observer class
class Observer
  attr_reader :last_event

  def update(event)
    @last_event = event
    puts "Observer received event: #{event}"
  end
end

# Iterator module
module Enumerable
  def each
    raise NotImplementedError
  end

  def map(&block)
    result = []
    each { |item| result << block.call(item) }
    result
  end

  def select(&block)
    result = []
    each { |item| result << item if block.call(item) }
    result
  end

  def reduce(initial, &block)
    accumulator = initial
    each { |item| accumulator = block.call(accumulator, item) }
    accumulator
  end
end

# Range class with Enumerable
class Range
  include Enumerable

  attr_reader :start, :finish

  def initialize(start_val, end_val)
    @start = start_val
    @finish = end_val
  end

  def each(&block)
    current = @start
    while current <= @finish
      block.call(current)
      current += 1
    end
  end

  def size
    @finish - @start + 1
  end
end

# Module constants
DEFAULT_BUFFER_SIZE = 4096
MAX_RETRIES = 3
VERSION = "2.0.0"

File: src/parser/__tests__/fixtures/csharp/sample2.cs
// sample2.cs - Interfaces, generics, LINQ, and advanced patterns

using System;
using System.Collections.Generic;
using System.Linq;

namespace AdvancedFeatures
{
    // Generic repository interface
    public interface IRepository<T> where T : class
    {
        T GetById(int id);
        IEnumerable<T> GetAll();
        void Add(T entity);
        void Update(T entity);
        void Delete(int id);
    }

    // Comparable interface
    public interface IComparable<T>
    {
        int CompareTo(T other);
    }

    // Drawable interface
    public interface IDrawable
    {
        void Draw();
        Rectangle GetBounds();
    }

    // Generic container class
    public class Container<T>
    {
        private T _value;

        public Container(T value)
        {
            _value = value;
        }

        public T Get()
        {
            return _value;
        }

        public void Set(T value)
        {
            _value = value;
        }

        public Container<TResult> Map<TResult>(Func<T, TResult> mapper)
        {
            return new Container<TResult>(mapper(_value));
        }

        public static Container<T> Wrap(T value)
        {
            return new Container<T>(value);
        }
    }

    // Generic pair class
    public struct Pair<TFirst, TSecond>
    {
        public TFirst First { get; set; }
        public TSecond Second { get; set; }

        public Pair(TFirst first, TSecond second)
        {
            First = first;
            Second = second;
        }

        public Pair<TSecond, TFirst> Swap()
        {
            return new Pair<TSecond, TFirst>(Second, First);
        }
    }

    // Result class for error handling
    public class Result<T>
    {
        public T Value { get; private set; }
        public Exception Error { get; private set; }

        private Result(T value, Exception error)
        {
            Value = value;
            Error = error;
        }

        public bool IsSuccess => Error == null;
        public bool IsFailure => !IsSuccess;

        public static Result<T> Success(T value)
        {
            return new Result<T>(value, null);
        }

        public static Result<T> Failure(Exception error)
        {
            return new Result<T>(default(T), error);
        }

        public T Unwrap()
        {
            if (IsFailure)
                throw Error;
            return Value;
        }
    }

    // Shape hierarchy with IDrawable
    public abstract class Shape : IDrawable
    {
        public string Color { get; set; }

        protected Shape(string color)
        {
            Color = color;
        }

        public abstract double Area();

        public virtual void Draw()
        {
            Console.WriteLine($"Drawing {GetType().Name}");
        }

        public abstract Rectangle GetBounds();
    }

    // Rectangle implementation
    public class Rectangle : Shape
    {
        public double Width { get; set; }
        public double Height { get; set; }

        public Rectangle(string color, double width, double height)
            : base(color)
        {
            Width = width;
            Height = height;
        }

        public override double Area()
        {
            return Width * Height;
        }

        public override Rectangle GetBounds()
        {
            return new Rectangle(Color, Width, Height);
        }
    }

    // Circle implementation
    public class CircleShape : Shape
    {
        public const double PI = 3.14159;

        public double Radius { get; set; }

        public CircleShape(string color, double radius)
            : base(color)
        {
            Radius = radius;
        }

        public override double Area()
        {
            return PI * Radius * Radius;
        }

        public override Rectangle GetBounds()
        {
            double diameter = 2 * Radius;
            return new Rectangle(Color, diameter, diameter);
        }
    }

    // Observer pattern implementation
    public interface IObserver<T>
    {
        void Update(T value);
    }

    public interface IObservable<T>
    {
        void Subscribe(IObserver<T> observer);
        void Unsubscribe(IObserver<T> observer);
        void Notify(T value);
    }

    // Subject implementation
    public class Subject<T> : IObservable<T>
    {
        private readonly List<IObserver<T>> _observers = new List<IObserver<T>>();
        private T _state;

        public T State
        {
            get => _state;
            set
            {
                _state = value;
                Notify(value);
            }
        }

        public void Subscribe(IObserver<T> observer)
        {
            if (!_observers.Contains(observer))
                _observers.Add(observer);
        }

        public void Unsubscribe(IObserver<T> observer)
        {
            _observers.Remove(observer);
        }

        public void Notify(T value)
        {
            foreach (var observer in _observers)
            {
                observer.Update(value);
            }
        }
    }

    // Observer implementation
    public class Observer<T> : IObserver<T>
    {
        public T LastValue { get; private set; }

        public void Update(T value)
        {
            LastValue = value;
            Console.WriteLine($"Observer received: {value}");
        }
    }

    // Strategy pattern
    public interface ISortStrategy<T>
    {
        IEnumerable<T> Sort(IEnumerable<T> items);
    }

    public class QuickSortStrategy<T> : ISortStrategy<T> where T : IComparable<T>
    {
        public IEnumerable<T> Sort(IEnumerable<T> items)
        {
            // Simplified QuickSort implementation
            var list = items.ToList();
            list.Sort();
            return list;
        }
    }

    // Repository implementation
    public class InMemoryRepository<T> : IRepository<T> where T : class
    {
        private readonly Dictionary<int, T> _storage = new Dictionary<int, T>();
        private int _nextId = 1;

        public T GetById(int id)
        {
            return _storage.ContainsKey(id) ? _storage[id] : null;
        }

        public IEnumerable<T> GetAll()
        {
            return _storage.Values;
        }

        public void Add(T entity)
        {
            _storage[_nextId++] = entity;
        }

        public void Update(T entity)
        {
            // Update logic here
        }

        public void Delete(int id)
        {
            _storage.Remove(id);
        }
    }

    // Event handling
    public class Button
    {
        public event EventHandler Click;
        public event EventHandler<MouseEventArgs> MouseMove;

        public void OnClick()
        {
            Click?.Invoke(this, EventArgs.Empty);
        }

        public void OnMouseMove(int x, int y)
        {
            MouseMove?.Invoke(this, new MouseEventArgs(x, y));
        }
    }

    public class MouseEventArgs : EventArgs
    {
        public int X { get; set; }
        public int Y { get; set; }

        public MouseEventArgs(int x, int y)
        {
            X = x;
            Y = y;
        }
    }

    // Delegate types
    public delegate void LogHandler(string message);
    public delegate T Factory<T>() where T : new();
    public delegate TResult Transformer<TIn, TResult>(TIn input);

    // Constants
    public static class Configuration
    {
        public const int DefaultBufferSize = 4096;
        public const int MaxRetries = 3;
        public const string Version = "2.0.0";
    }
}
